phpBB/phpbb/install/module/update_database/task/update_extensions.php

```diff
@@ -0,0 +1,244 @@
+	 *				they were added
+	 */
+	static public $default_extensions_update = [
+		'3.2.0-b3' => ['phpbb/viglink']
+	];
+
+	/**
+	 * Constructor
+	 *
+	 * @param container_factory			$container
+	 * @param config					$install_config
+	 * @param iohandler_interface		$iohandler
+	 * @param $update_helper			$update_helper
+	 * @param string					$phpbb_root_path phpBB root path
+	 */
+	public function __construct(container_factory $container, config $install_config, iohandler_interface $iohandler, update_helper $update_helper, $phpbb_root_path)
+	{
+		$this->install_config	= $install_config;
+		$this->iohandler		= $iohandler;
+		$this->extension_table = $container->get_parameter('tables.ext');
+
+		$this->log				= $container->get('log');
+		$this->user				= $container->get('user');
+		$this->extension_manager = $container->get('ext.manager');
+		$this->cache				= $container->get('cache.driver');
+		$this->config			= $container->get('config');
+		$this->db				= $container->get('dbal.conn');
+		$this->update_helper = $update_helper;
+		$this->finder = new Finder();
+		$this->finder->in($phpbb_root_path . 'ext/')
+			->ignoreUnreadableDirs()
+			->depth('< 3')
+			->files()
+			->name('composer.json');
+
+		// Make sure asset version exists in config. Otherwise we might try to
+		// insert the assets_version setting into the database and cause a
+		// duplicate entry error.
+		if (!isset($this->config['assets_version']))
+		{
+			$this->config['assets_version'] = 0;
+		}
+
+		parent::__construct(true);
+	}
+
+	/**
+	 * {@inheritdoc}
+	 */
+	public function run()
+	{
+		$this->user->session_begin();
+		$this->user->setup(array('common', 'acp/common', 'cli'));
+
+		$update_info = $this->install_config->get('update_info_unprocessed', []);
+		$version_from = !empty($update_info) ? $update_info['version']['from'] : $this->config['version_update_from'];
+
+		if (!empty($version_from))
+		{
+			$update_extensions = $this->iohandler->get_input('update-extensions', []);
+
+			// Create list of default extensions that need to be enabled in update
+			$default_update_extensions = [];
+			foreach (self::$default_extensions_update as $version => $extensions)
+			{
+				if ($this->update_helper->phpbb_version_compare($version_from, $version, '<='))
+				{
+					$default_update_extensions = array_merge($default_update_extensions, $extensions);
+				}
+			}
+
+			$available_extensions = $this->extension_manager->all_available();
+
+			// Update available extensions
+			foreach ($available_extensions as $ext_name => $ext_path)
+			{
+				// Update extensions if:
+				//	1) Extension is currently enabled
+				//	2) Extension was implicitly defined as needing an update
+				//	3) Extension was newly added as default phpBB extension in
+				//		this update and should be enabled by default.
+				if ($this->extension_manager->is_enabled($ext_name) ||
+					in_array($ext_name, $update_extensions) ||
+					in_array($ext_name, $default_update_extensions)
+				)
```

Comment:
Looks so much cleaner now! :)
Anyway, since some administrators tend to install as many extensions as possible it might be cleverer to handle timeouts. I'm not sure how fast disabling and enabling an extension is without any migrations changes (which will be the case for most of the extensions during an update) but this is probably just an extreme case.